File: 0000-0999/0410.SplitArrayLargestSum.go
func splitArray(nums []int, m int) int {
    left, right := 0, 0
    for _, num := range nums {
        left = max(left, num)
        right += num
    }
    for left < right {
        mid := (left+right) >> 1
        if split(nums, mid) <= m {
            right = mid
        } else {
            left = mid + 1
        }
    }
    return left
}

func split(nums []int, target int) int {
    cnt, sum := 1, 0
    for _, num := range nums {
        sum += num
        if sum > target {
            cnt++
            sum = num
        }
    }
    return cnt
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: 0000-0999/0410.SplitArrayLargestSum.py
class Solution(object):
    def splitArray(self, nums, m):
        """
        :type nums: List[int]
        :type m: int
        :rtype: int
        """
        low, high = max(nums), sum(nums)
        while low < high:
            mid = low + (high-low)/2
            if self.isVaild(nums, m, mid):
                high = mid
            else:
                low = mid + 1
        return high
        
    def isVaild(self, nums, m, test):
        cur = 0
        count = 1
        for num in nums:
            cur += num
            if cur > test:
                cur = num
                count += 1
                if count > m:
                    return False
        return True
        